### PR TITLE
Feature/validate metric name uniqueness

### PR DESCRIPTION
**Pull request by ycabrer**
Metric Names are required to be unique in Scaled Objects

### Checklist

- [X] Commits are signed with Developer Certificate of Origin (DCO)
- [X] Tests have been added
- [N/A] A PR is opened to update the documentation on https://github.com/kedacore/keda-docs
- [X] Changelog has been updated

Fixes #

### CODE REVIEW

**Comment by ycabrer** [comment]
@zroubalik I'm not sure why the static check is failing. It seems to be referencing code in the object finilizers. I didn't modify them (I think).

Also any tips on where to put the documentation update for the metric name validation?

File: controllers/scaledobject_controller.go [hunk 1]
```diff
@@ -214,6 +219,29 @@ func (r *ScaledObjectReconciler) ensureScaledObjectLabel(logger logr.Logger, sca
...
+func (r *ScaledObjectReconciler) validateMetricNameUniqueness(logger logr.Logger, scaledObject *kedav1alpha1.ScaledObject) error {
+	scalers, err := r.scaleHandler.GetScalers(scaledObject)
+	if err != nil {
+		return err
+	}
+
+	observedMetricNames := make(map[string]struct{})
+	for _, scaler := range scalers {
+		for _, metric := range scaler.GetMetricSpecForScaling() {
+			metricName := metric.External.Metric.Name
```
Comment (zroubalik) [inline controllers/scaledobject_controller.go:231]: We should check whether the scaler contains External metrics (`if != nil`), becasue if Resource based Metrics are used, External Metrics are not set.

See: https://github.com/kedacore/keda/blob/af2b36f593c231ed25c48a0ba432f6128f77cbf5/controllers/hpa.go#L135-L162

**Review by zroubalik — CHANGES_REQUESTED** [review]
Looking good so far, we should add a check though

**Comment by zroubalik** [comment]
> @zroubalik I'm not sure why the static check is failing. It seems to be referencing code in the object finilizers. I didn't modify them (I think).

The static check is strange, you can probably disable this specific one for the finalizers: https://github.com/kedacore/keda/blob/af2b36f593c231ed25c48a0ba432f6128f77cbf5/.golangci.yml#L50

> 
> Also any tips on where to put the documentation update for the metric name validation?

We should probably add a new section unde [KEDA Concepts](https://keda.sh/docs/2.0/concepts/), something like section `Mutliple Scalers` or `Using Scalers` ? @tomkerkhove WDYT?

**Comment by zroubalik** [comment]
> > @zroubalik I'm not sure why the static check is failing. It seems to be referencing code in the object finilizers. I didn't modify them (I think).
> 
> The static check is strange, you can probably disable this specific one for the finalizers:
> 
> https://github.com/kedacore/keda/blob/af2b36f593c231ed25c48a0ba432f6128f77cbf5/.golangci.yml#L50
> 
> > Also any tips on where to put the documentation update for the metric name validation?
> 
> We should probably add a new section unde [KEDA Concepts](https://keda.sh/docs/2.0/concepts/), something like section `Mutliple Scalers` or `Using Scalers` ? @tomkerkhove WDYT?

@ycabrer any update please? :)

**Comment by tomkerkhove** [comment]
> > Also any tips on where to put the documentation update for the metric name validation?
> 
> We should probably add a new section unde [KEDA Concepts](https://keda.sh/docs/2.0/concepts/), something like section `Mutliple Scalers` or `Using Scalers` ? @tomkerkhove WDYT?

Missed this one, sorry @zroubalik, but I agree to have a section on multiple scalers and what to expect of them.

Not sure though if this should include technical details such as metric name validation but more how the decision making will happen maybe?

**Comment by zroubalik** [comment]
@ycabrer are you able to update this please? It is completely OK, if you don't have a time to do so. But please let us know, so we can manage this. Thanks!

**Comment by ycabrer** [comment]
@zroubalik Sorry for the delay. I've had a busy start to the new year! I'll go ahead and update the docs.

File: controllers/scaledobject_controller.go (hunk 1 above)
Comment (ycabrer) [reply controllers/scaledobject_controller.go]: That's a great point! I went ahead and added a continue when external metrics is nil

**Comment by ycabrer** [comment]
> > > @zroubalik I'm not sure why the static check is failing. It seems to be referencing code in the object finilizers. I didn't modify them (I think).
> > 
> > 
> > The static check is strange, you can probably disable this specific one for the finalizers:
> > https://github.com/kedacore/keda/blob/af2b36f593c231ed25c48a0ba432f6128f77cbf5/.golangci.yml#L50
> > 
> > > Also any tips on where to put the documentation update for the metric name validation?
> > 
> > 
> > We should probably add a new section unde [KEDA Concepts](https://keda.sh/docs/2.0/concepts/), something like section `Mutliple Scalers` or `Using Scalers` ? @tomkerkhove WDYT?
> 
> @ycabrer any update please? :)

I rebased my branch against kedacore/main. It seems like the static checks are still failing for me on the finalizers. I took a look at the golang ci config file and can't seem to ignore the failures. Any advice?

**Comment by zroubalik** [comment]
> > > > @zroubalik I'm not sure why the static check is failing. It seems to be referencing code in the object finilizers. I didn't modify them (I think).
> > > 
> > > 
> > > The static check is strange, you can probably disable this specific one for the finalizers:
> > > https://github.com/kedacore/keda/blob/af2b36f593c231ed25c48a0ba432f6128f77cbf5/.golangci.yml#L50
> > > 
> > > > Also any tips on where to put the documentation update for the metric name validation?
> > > 
> > > 
> > > We should probably add a new section unde [KEDA Concepts](https://keda.sh/docs/2.0/concepts/), something like section `Mutliple Scalers` or `Using Scalers` ? @tomkerkhove WDYT?
> > 
> > 
> > @ycabrer any update please? :)
> 
> I rebased my branch against kedacore/main. It seems like the static checks are still failing for me on the finalizers. I took a look at the golang ci config file and can't seem to ignore the failures. Any advice?

Have you tried adding a new section to the `excluded-rules` section in `.golangci.yaml`?

Something like this:
```yaml
# Excluding interfacer for finalizers, reason: https://github.com/kedacore/keda/pull/1390
 - path: _finalizer.go
      linters:
        - interfacer
```

**Comment by ycabrer** [comment]
> > > > > @zroubalik I'm not sure why the static check is failing. It seems to be referencing code in the object finilizers. I didn't modify them (I think).
> > > > 
> > > > 
> > > > The static check is strange, you can probably disable this specific one for the finalizers:
> > > > https://github.com/kedacore/keda/blob/af2b36f593c231ed25c48a0ba432f6128f77cbf5/.golangci.yml#L50
> > > > 
> > > > > Also any tips on where to put the documentation update for the metric name validation?
> > > > 
> > > > 
> > > > We should probably add a new section unde [KEDA Concepts](https://keda.sh/docs/2.0/concepts/), something like section `Mutliple Scalers` or `Using Scalers` ? @tomkerkhove WDYT?
> > > 
> > > 
> > > @ycabrer any update please? :)
> > 
> > 
> > I rebased my branch against kedacore/main. It seems like the static checks are still failing for me on the finalizers. I took a look at the golang ci config file and can't seem to ignore the failures. Any advice?
> 
> Have you tried adding a new section to the `exclude-rules` section in `.golangci.yaml`?
> 
> Something like this:
> 
> ```yaml
> # Excluding interfacer for finalizers, reason: https://github.com/kedacore/keda/pull/1390
>  - path: _finalizer.go
>       linters:
>         - interfacer
> ```

That worked 

> > > > > @zroubalik I'm not sure why the static check is failing. It seems to be referencing code in the object finilizers. I didn't modify them (I think).
> > > > 
> > > > 
> > > > The static check is strange, you can probably disable this specific one for the finalizers:
> > > > https://github.com/kedacore/keda/blob/af2b36f593c231ed25c48a0ba432f6128f77cbf5/.golangci.yml#L50
> > > > 
> > > > > Also any tips on where to put the documentation update for the metric name validation?
> > > > 
> > > > 
> > > > We should probably add a new section unde [KEDA Concepts](https://keda.sh/docs/2.0/concepts/), something like section `Mutliple Scalers` or `Using Scalers` ? @tomkerkhove WDYT?
> > > 
> > > 
> > > @ycabrer any update please? :)
> > 
> > 
> > I rebased my branch against kedacore/main. It seems like the static checks are still failing for me on the finalizers. I took a look at the golang ci config file and can't seem to ignore the failures. Any advice?
> 
> Have you tried adding a new section to the `exclude-rules` section in `.golangci.yaml`?
> 
> Something like this:
> 
> ```yaml
> # Excluding interfacer for finalizers, reason: https://github.com/kedacore/keda/pull/1390
>  - path: _finalizer.go
>       linters:
>         - interfacer
> ```

That did it! I had chosen the wrong linter.

File: .golangci.yml [hunk 2]
```diff
@@ -56,6 +56,9 @@ issues:
     - path: scale_handler.go
       linters:
         - gocyclo
+    - path: _finalizer.go
```
Comment (zroubalik) [inline .golangci.yml:59]: Could you please add a comment with a description why are we adding this exception, eg. with a link to this PR?

File: CHANGELOG.md [hunk 3]
```diff
@@ -40,6 +40,7 @@
 - Improve error reporting in prometheus scaler ([PR #1497](https://github.com/kedacore/keda/pull/1497))
 
 ### Breaking Changes
+- Require metricNames be unique in scaled objects.
```
Comment (zroubalik) [inline CHANGELOG.md:43]:
```suggestion
- Require metricNames be unique in scaled objects ([#1390](https://github.com/kedacore/keda/pull/1390))
```

File: .golangci.yml (hunk 2 above)
Comment (ycabrer) [reply .golangci.yml:60]: Sure thing! I missed that you had already given an example

File: CHANGELOG.md (hunk 3 above)
Comment (ycabrer) [reply CHANGELOG.md]: I just updated it

**Review by zroubalik — APPROVED** [review]
LGTM, thanks @ycabrer!

**Pull request merged**